### PR TITLE
Improve imperative mode training speed

### PR DESCRIPTION
**Pull request by velconia**
1. Accelerate mnist according the profiling result.
2. Add binutils 2.27 to manylinux dockerfile.
3. Implement gperf profiler in imperative mode.

### CODE REVIEW

File: paddle/fluid/imperative/profiler.cc [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+#else
+      LOG(WARNING) << "Paddle is not compiled with gperftools. "
+                      "FLAGS_tracer_profile_fname will be ignored";
+#endif
+    });
+  }
+}
+
+void StopProfile() {
+  LOG(ERROR) << "ZZ " << FLAGS_tracer_profile_fname;
```
Comment (panyx0718) [inline paddle/fluid/imperative/profiler.cc]: ?

File: paddle/fluid/imperative/profiler.cc [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+#endif
+
+void StartProfile() {
+  LOG(ERROR) << "XX " << FLAGS_tracer_profile_fname;
+  if (!FLAGS_tracer_profile_fname.empty()) {
+    std::call_once(gTracerProfileOnce, [] {
+#ifdef WITH_GPERFTOOLS
+      ProfilerStart(FLAGS_tracer_profile_fname.c_str());
+      gTracerProfilerStarted = true;
+      LOG(ERROR) << "YY";
```
Comment (panyx0718) [inline paddle/fluid/imperative/profiler.cc]: ?

File: paddle/fluid/framework/grad_op_desc_maker.h [hunk 3]
```diff
@@ -55,7 +55,10 @@ class GradOpDescMakerBase {
                    std::back_inserter(ret_val),
                    [this](const std::string& fwd_var_name) -> std::string {
                      auto g_name = GradVarName(fwd_var_name);
-                     if (no_grad_set_.count(g_name)) {
+                     if (no_grad_set_.empty()) {
```
Comment (panyx0718) [inline paddle/fluid/framework/grad_op_desc_maker.h]: `no_grad_set_.empty() || !no_grad_set_.count(g_name)`

and remove the else below?

File: paddle/fluid/imperative/profiler.cc (hunk 1 above)
Comment (velconia) [reply paddle/fluid/imperative/profiler.cc]: Done

File: paddle/fluid/imperative/profiler.cc (hunk 2 above)
Comment (velconia) [reply paddle/fluid/imperative/profiler.cc]: Done

File: paddle/fluid/framework/grad_op_desc_maker.h (hunk 3 above)
Comment (velconia) [reply paddle/fluid/framework/grad_op_desc_maker.h]: Done

**Review by panyx0718 — APPROVED** [review]

**Pull request merged**